### ISSUE TITLE
Consider making the `src` cache read-only.

### ISSUE THREAD

**Issue by ehuss** [rust-lang/cargo#9455]
Registry dependencies get extracted into cargo home (in the `src` directory) with whatever metadata is in the tar file.  One issue with this is that the files are usually writeable. This can cause a problem if the user accidentally modifies these files, which breaks cargo's assumption that they are immutable and reusable. One way this can happen is that in some editors, when there is an error or warning, they may open those files to display the error/warning (particularly with macros).  The user may not realize that this is from a remote location, and may not understand the consequence of making changes.

We may want to consider making those files read-only when extracting them. This would help with confusing situations where the `src` cache is inadvertently corrupted.

This would not protect from general filesystem corruption, which is also an issue. This is also an issue with `git` dependencies, which may be more difficult to adjust permissions on.

There is some risk that this would introduce new problems. For example, if people are using tools to clean the src directory, and those tools have trouble with read-only files.

**Comment by matthiaskrgr** [comment]
Oh that's an interesting idea!
While I fully agree that it makes sense to make the cache (or at least `.cargo/registry/src` and `.cargo/git/checkout`) write-only, it does indeed prevent cargo-cache from cleaning up anything;

````
$ cargo cache --autoclean
Clearing cache...

Warning: failed to recursively remove directory "/home/matthias/.cargo/registry/src".
error: Os { code: 13, kind: PermissionDenied, message: "Permission denied" }
Warning: failed to recursively remove directory "/home/matthias/.cargo/git/checkouts".
error: Os { code: 13, kind: PermissionDenied, message: "Permission denied" }
````

BUT I am already happy that it does not panic :D
If it turned out that `cargo-cache` does modify the cache in any way (anything that goes beyond just removing whole files) I would indeed consider this a bug.

Right now I am just running `remove_dir_all::remove_dir_all()` on directories without doing any checks whatsoever.
https://crates.io/crates/remove_dir_all
I never had a read-only cache in mind when writing `cargo-cache` but adding a little bit of extra logic which tries to change  file permission to read-write before removing them does not sound too difficult, assuming that we have a crate around that handles file permissions on linux, macos and windows.

**Comment by eminence** [comment]
I'm bringing up a new rust user on my team, and I found that they ended up editing some files in $CARGO_HOME  (which is actually quite easy to do with VSCode+RA) to add some debugging messages.   A read-only src cache might have given them a sufficient hint that what they were doing was not the right approach.

**Comment by weihanglo** [comment]
Just [did a PoC] and benchmarks with rust-lang/cargo itself. It turns out that there is a slight performance hit (~5%) on macOS. It's in expectation, as there are some extra permission operations on every file for the first time checking out/unpacking a crate (Windows might do it better since file permissions can be inherited, though). 

However, it does get **new problems** 🤯. When I removed sources residing in `~/.cargo/registry/src` and re-run the build again, I got a permission denied error. 

```console
$ ./target/release/cargo b
$ rm -rf ~/.cargo/registry/src
$ ./target/release/cargo b
error: failed to run custom build command for `curl-sys v0.4.55+curl-7.83.1`

Caused by:
  process didn't exit successfully: `/Users/myuser/projects/cargo/target/debug/build/curl-sys-f4af01668b6e1680/build-script-build` (exit status: 101)
  --- stdout
  ...
  --- stderr
  fatal: not a git repository (or any of the parent directories): .git
  thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 13, kind: PermissionDenied, message: "Permission denied" }', /Users/myuser/.cargo/registry/src/github.com-1ecc6299db9ec823/curl-sys-0.4.55+curl-7.83.1/build.rs:93:10
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I guess its curl-sys's build script that uses `fs::copy` to copy contents from `registry/src` to build script's `OUT_DIR`, but unfortunately `fs::copy` also copies permission bits. Any rebuild that triggers a re-run of the same build-script binary might fail because it cannot `fs::copy` from cache into existing **readonly** files under target-dir. I am afraid that more cases in the wild will be broken if `registry/src` becomes readonly. Here is [the search result] that a build.rs contains `fs::copy` throughout GitHub.

Personally, I love to see this enhancement landed, but it seems a bit risky to me after the analysis.

[did a PoC]: https://github.com/rust-lang/cargo/commit/b7be2609cc2e763650c24ccd24d5c5d61aa8a12a
[the search result]: https://github.com/search?q=filename%3Abuild.rs+language%3ARust+%22fs%3A%3Acopy%22&type=Code

**Comment by ehuss** [comment]
That's really unfortunate, I hadn't thought about that. 

I'm wondering if a compromise would be to only adjust permissions on packages that don't have a build script?  I don't know if it would be easy to detect that (that might require parsing Cargo.toml, which might be too much overhead).  I also don't know what % of packages that would cover.

It's also unfortunate that the `tar` API doesn't allow us to override the permissions directly.  Another option is to make the files readonly when creating the `.crate` file. That would only help with newly published packages, but might be an option to consider.

**Comment by nagisa** [comment]
Given that most of these cases would be caught by crater, maybe its worthwhile to just make the change and make sure crates are fixed up?

Alternatively: maybe only apply this for crates published after a certain timestamp?

**Comment by ehuss** [comment]
@weihanglo Asked about how to do a crater run to test the impact of changing the readonly status. The rough steps are:

1. In your clone of cargo, make the changes to incorporate the new behavior.
2. Get a clone of https://github.com/rust-lang/rust/
3. Check out a branch to add your changes.
4. Modify `.gitmodules` to point to your clone and branch of cargo with the changes you want to test.  For example:
   ```bash
   git submodule set-url src/tools/cargo https://github.com/ehuss/cargo.git
   git submodule set-branch --branch my-awesome-feature src/tools/cargo
   git submodule update --remote src/tools/cargo
   git add .gitmodules src/tools/cargo
   git commit
   ```
5. Push the changes to GitHub and make a PR . Clearly label the PR as an experiment, and assign yourself or @ghost.
6. Make a "try" build. I think all cargo members have this permission. Write a comment `@bors try`.
7. After the try build finishes (an hour? I forget), ask someone to make a crater run. The Cargo team does not have that permission, so just as someone on Zulip.

I think doing a "check" run should be sufficient?  I'm concerned that by default it will not get dev-dependencies, so maybe add the `--all-targets` flag? I'm not sure if that's worth it. There's a bunch of documentation at https://github.com/rust-lang/crater/blob/master/docs/bot-usage.md on crater commands.  You may also consider doing a top-100 run first just to make sure the experiment is configured properly.

----

I still feel that it might be good to ease into the change, such as only marking read-only if there are no build scripts. We discussed some other options to consider:
* Adding some kind of option so that build-scripts can say "don't mark me read-only" or "don't mark these directories read-only".
* Make all packages read-only on an edition boundary. This wouldn't be able to support automatic migrations, which conflicts with the spirit of Editions being easy to transition.

Doing a crater run can maybe give us a sense of the scale of the problem.

**Comment by kornelski** [comment]
This could be limited to only `*.rs` files, leaving other data files with a writeable bit.

**Comment by bjorn3** [comment]
Cargo.{toml,lock} should probably also be read-only.

**Comment by torhovland** [comment]
As suggested [here](https://internals.rust-lang.org/t/cargo-should-detect-changes-in-dependency-source-files), a possible alternative is to compute a hash when downloading a dependency, and to verify that hash whenever Cargo needs to access that dependency.

**Comment by matklad** [comment]
Somewhat related, we might consider sanitizing the executable bit as well, if we consider distribution of compiled executables via crate tarballs an unsupported use-case.

**Comment by wizeguyy** [comment]
If these files are made readonly, what would the process be for updating dependencies in the cache? i.e. if `cargo update` detects a new version of a crate in the cache, it will need write permissions to update that crate.

Will cargo temporarily modify the file permissions before performing an update? On unix systems it might make more sense for cargo to run as its own user. Other users have read-only access, but the cargo user has write access. Thoughts?

**Comment by bjorn3** [comment]
Every crate version is stored in a separate directory. This is necessary to be able to cache multiple crate versions at once. And because of this we never need to modify the directory containing the cached crate after we are done unpacking the downloaded tarball.

**Comment by bjorn3** [comment]
> On unix systems it might make more sense for cargo to run as its own user.

There is no way to do that as an unprivileged user. And requiring root permission to install rust would be a really bad idea.

**Comment by wizeguyy** [comment]
Ahh... I should have seen that. In that case, disregard my comment. In that case, I am fully in support of making the registry src files read-only.

**Comment by sanmai-NL** [comment]
We can in addition or instead of ownership or permission bits, use OS-specific flags for finer-grained semantics, like [`chattr` to make files immutable](https://en.wikipedia.org/wiki/Chattr#Attributes) under Linux. Packages that try to break these restrictions will often be detected and the fixes will benefit even users of other operating systems.

**Comment by weihanglo** [comment]
@sanmai-NL could you elaborate more on how the approach resolves the concern in <https://github.com/rust-lang/cargo/issues/9455#issuecomment-1143026498>? It may still fail people's builds, no?

**Comment by bjorn3** [comment]
The immutable flag on linux can only be set by root as it prevents even root from modifying it.

**Comment by sanmai-NL** [comment]
@weihanglo Better performance can be attained by setting the right inode metadata along with file creation. Can you test setting an umask to make newly created files by Cargo read-only by default? The immutable bit tip was more of an implementation enhancement suggestion on attributes other than performance.

As for the breakage, I think that it's a defect to perform these file manipulations in build scripts.

I don't know if it is possible within the Cargo architecture to restrict build scripts to not perform some actions. That would be good in general.

**Comment by bjorn3** [comment]
The umask is process global, but someone may be creating files concurrently with cargo when using cargo as library. Also for directories setting the umask I believe will require using chmod anyway to make them writable again while creating files inside the directory. And on Unix I'm not too worried about a single extra chmod per file and directory. On Windows I would be more concerned, but the Windows kernel doesn't have anything like umask afaik. The CRT supports the function, but I wouldn't be surprised if it internally calls the equivalent of chmod after every file creation.

**Comment by kornelski** [comment]
This could be opt-in per crate. Cargo currently makes tarballs with `-rw-r--r--` permissions (ignores original files' permissions). Cargo could be taught to make tarballs with `-r--r--r--` permissions, and preserve that when unarchiving.

**Comment by sanmai-NL** [comment]
Just an idea here off the top of my head, but why extract source archives at all? Can't we get their contents as needed, in-memory? That may sound costly but I wonder how often that really happens and if extracting them amortizes well (the extra cost being the filesystem operations that notably result in syscalls) That would reduce the number of files to be created and changed, so even particular file metadata like immutable flag can be set at the cost of $O(n)$ instead of $O(nm)$ time in the average case, where $n$ is the number of crates (tar archives) and $m$ is the number of files per crate.

**Comment by sanmai-NL** [comment]
> This could be opt-in per crate. Cargo currently makes tarballs with `-rw-r--r--` permissions (ignores original files' permissions). Cargo could be taught to make tarballs with `-r--r--r--` permissions, and preserve that when unarchiving.

@kornelski 

https://docs.rs/tar/latest/tar/struct.Entry.html#method.set_mask

**Comment by bjorn3** [comment]
> Can't we get their contents as needed, in-memory?

Rustc currently doesn't support that. We could add it, but build scripts and proc macros will still need it extracted somewhere on the filesystem.

**Comment by mathstuf** [comment]
> Can't we get their contents as needed, in-memory?

Tools communicate by filesystem paths, so how do you propose to send an in-memory "path" to `sccache` for caching? Or coverage tools that render the coverage data with the source code? That's even if we do manage to ignore `build.rs` scripts that do…whatever they do. Without some cross-platform FUSE-like solution that exposes such paths to arbitrary other tools, this would restrict the current set of use cases. Maybe that's fine and can be opt-in, but opt-out is, I believe, only going to continue finding more and more corner cases.

**Comment by kornelski** [comment]
The invention of not-a-tarball crate storage system is irrelevant here, because the problem is crates relying on files existing in the file system, and existing with specific permissions. If you succeed in removing the files from `src` dir, you'll break them even worse.

**Comment by sanmai-NL** [comment]
> > Can't we get their contents as needed, in-memory?
> 
> Tools communicate by filesystem paths, so how do you propose to send an in-memory "path" to `sccache` for caching? Or coverage tools that render the coverage data with the source code? That's even if we do manage to ignore `build.rs` scripts that do…whatever they do. Without some cross-platform FUSE-like solution that exposes such paths to arbitrary other tools, this would restrict the current set of use cases. Maybe that's fine and can be opt-in, but opt-out is, I believe, only going to continue finding more and more corner cases.

Does `sccache` cache source code and would that be important for Rust? The challenge would be more to let `rustc` compile tar archives, no?

**Comment by sanmai-NL** [comment]
> The invention of not-a-tarball crate storage system is irrelevant here, because the problem is crates relying on files existing in the file system, and existing with specific permissions. If you succeed in removing the files from `src` dir, you'll break them even worse.

What is your opinion on this breakage? Would it preclude moving forward with a more performant design in any case?

**Comment by bjorn3** [comment]
Sccache doesn't cache source code. It reads the `.d` file emitted by rustc to know when it needs to recompile. If rustc were to compile tarballs, this `.d` file would simply mention the tarball and thus sccache would work as expected. For something entirely in-memory passed in through eg an fd or unix socket, that wouldn't work though.

**Comment by sanmai-NL** [comment]
Let's also discuss differentation between crates. With our without build scripts where mentioned. But only-binary or only-library is also relevant. Binary-only crates's source as some specific version needs to be only read very infrequently compared to libraries, especially popular libraries. So for binary-only, non-buildscript crates you'd most want to avoid extracting the tar archive (even though they may be smaller, in comparison, perhaps).

**Comment by sanmai-NL** [comment]
> Sccache doesn't cache source code. It reads the `.d` file emitted by rustc to know when it needs to recompile. If rustc were to compile tarballs, this `.d` file would simply mention the tarball and thus sccache would work as expected. For something entirely in-memory passed in through eg an fd or unix socket, that wouldn't work though.

Yes, and the latter would probably require much more significant architecture reengineering than supporting tar-like archives directly.

**Comment by sanmai-NL** [comment]
> Let's also discuss differentation between crates. Crates without build scripts were mentioned. But only-binary or only-library is also relevant. Binary-only crates's source as some specific version needs to be only read very infrequently compared to libraries, especially popular libraries. So for binary-only, non-buildscript crates you'd most want to avoid extracting the tar archive (even though they may be smaller, in comparison, perhaps) and perhaps doing other filesystem operations such as setting permissions.

And this assuming that no incremental compilation cache is available. In my experience one sometimes cleans this cache. But not too often I suppose.

**Comment by kornelski** [comment]
Can this be done for 2024 edition crates?

`Cargo.toml` is usually at the very beginning of tarballs, so it's relatively cheap to check edition of a `.crate` file before decompressing it for real to the file system.

**Comment by epage** [comment]
Independent of the question of whether this can be tied to an Edition, they asked for Edition 2024 changes to have their PRs posted by May 1st.  There is a slim chance something small can be rushed in late but members of the Cargo team have already been exhausting themselves to meet the May 1st date with what was already committed.

**Comment by HKalbasi** [comment]
An alternative to making the registry read only is to detect changes (for example by looking at `mtime`) and warn user and download/unpack the crate again. Would it solve the problem?

**Comment by bjorn3** [comment]
That would be slower as you did have to check it on every build rather than set the readonly flag once, right?

**Comment by HKalbasi** [comment]
I think we can just check the `mtime` of the root directory, and check `mtime` of individual crates only when the root one is wrong. So it would be a `O(1)` check in the happy path.

**Comment by bjorn3** [comment]
The mtime of the root directory is not updated when a file inside a subdirectory changes.

**Comment by kornelski** [comment]
The check could be asynchronous, running in parallel with the build.

**Comment by epage** [comment]
Going back a bit to the discussion for the PoC

> I'm wondering if a compromise would be to only adjust permissions on packages that don't have a build script? I don't know if it would be easy to detect that (that might require parsing Cargo.toml, which might be too much overhead). I also don't know what % of packages that would cover.

With #13713, we can parse `Cargo.toml` and check if `package.build = false` and mark them read-only in that case.  We don't even need to do a full manifest parse (including target discovery), just use `cargo-util-schemas` or even a one-off type for accessing that field (which carries risk of getting out of sync with the rest of the code).

Considering this is an extraction-time cost, I suspect the overhead for that minimal parse would be acceptable.  If it is a high cost, we could put it in the Index.

This won't help with packages published with a Cargo older than 1.80 but over time the coverage will improve.  We'd already be adopting a limited approach anyways as packages with build scripts would be skipped though #14948 has ways to help reduce the presence of those.

**Comment by epage** [comment]
If we add a check, I would expect it would be in a health check command, like #9341.

**Comment by epage** [comment]
Inspired by #6008, what if we mark all files as read-only when creating the `.crate` file if there is no build script?

**Comment by mqudsi** [comment]
I ran into this [in a very ugly fashion](https://github.com/tokio-rs/axum/issues/3404) last month; presumably I jumped to definition in vim and then an errant mouse or keyboard click pasted valid, 100% valid & compiling code in place of the original trait definition, leading me on a wild goose chase that wasted considerable time. I feel stupid admitting what happened but also feel like it's a great learning opportunity: this stuff happens and it really sucks when it does.

I ended up here after searching for existing issues regarding checking out a readonly copy of the registry and, at the time, didn't have anything to add after reading the very valid conflicting concerns between correctness and performance. However, I just recently had some thoughts that might perhaps inspire at least a mitigation if not a full solution.

* If using freshness (presumably a more recent mtime for a source file than what you would expect?), there is a compromise between making it an opt-in and forcibly enabling it everywhere at the cost of a non-negligible performance hit: you can keep the opt-in for maximum correctness, but default to only check the source mtime in case of a compilation failure (*possibly* even limiting it only to source files directly contributing to the compilation unit that is failing, rather than the entire dependency graph). This keeps the happy case happy, CI won't see a hit in the default case, but you can still trace issues caused by corrupted/modified copies of checked out sources.
* This one is rather complicated and probably not worth pursuing but if you integrate modified sources logic deep enough in the stack, you can pick it up as an unexpected source of dependency invalidation; i.e. checksum/mtime of individual source file(s) shouldn't be the reason for rebuilding a project dependency unless the corresponding package hash/git revision changed.
* While it's true that on \*nix parent folder mtime isn't (normally) touched when children are modified and that read-only file modes do not propagate to children (so you can't just RO the entire directory), the big loop hole is that *ownership* does propagate. I know cargo operates with user privileges, but hypothetically if the owner/group of the cargo registry were a different user, write privileges could be denied to the user while still allowing cargo to modify registry contents at will.
* In the same vein and while absolutely no one likes to talk about ACL and \*nix because it's a mess and depends on a lot of different variables/factors, it's probably possible (on a subset of supported machines) to use ACL to hack around the non-recursive RO limitation while still allowing cargo to operate as the same user (probably via O(1) hacks such as removing and then re-applying the ACL rule when it's time to update the registry).
* Getting into the "half-baked solution but better than no solution at all" territory: the rust extensions for the most popular editors (VSCode, (neo)vim, etc) could set, in the editor upon load of a buffer sourced from the registry checkout, an editor-specific, transient readonly flag to prevent the user from saving the changes without a forcible write (either "save anyway" or `:w!`) This would actually probably be the fastest and easiest stop-gap solution; the following actually works great for my biggest needs and concerns:
```vim
" Prevent accidentally modifying cargo registry
autocmd BufRead,BufNewFile,FileReadPost,FilterReadPost
    \ */.cargo/registry/src/* setlocal readonly
```

**Comment by bjorn3** [comment]
Rust-analyzer should already set the readonly flag when goto-definition jumping to a non-local dependency at least in vscode. I don't know if it works in other editors too.

**Comment by mqudsi** [comment]
> Rust-analyzer should already set the readonly flag when goto-definition jumping to a non-local dependency at least in vscode. I don't know if it works in other editors too.

I just tested it in neovim with the default lsp-client and rust-analyzer nightly and it's definitely not the case here. Works great with the `autocmd` I posted above, though.

**Comment by mathstuf** [comment]
The `autocmd` also works if you end up there by other means too (e.g., `locate` or `ripgrep`).

**Comment by kornelski** [comment]
The LSP protocol does not support read-only flag. Rust-analyzer implements extra features and workarounds for the Microsoft's editor, but does not support other editors beyond the LSP protocol.

Sublime Text has the same problem with rewriting files, and it's a WONTFIX from rust-analyzer.

**Comment by mqudsi** [comment]
Thanks for that info, @kornelski.

Just for clarity/posterity, my recommendation to use the autocmd (or whatever editor-specific equivalent) was not intended to be performed at the LSP level but at the "default editor plugin (even if upstream) for basic rust support" level. I know `rust.vim` is abandonware, but imagine if it were replaced with a much slimmer `rust.vim` that basically sets up the lsp integration, does things like the autocmd I mentioned, etc (or this can be the official request from the `rust-lang` team to the (neo)vim maintainers for the upstream filetype plugin).

**Comment by Skgland** [comment]
> The LSP protocol does not support read-only flag. Rust-analyzer implements extra features and workarounds for the Microsoft's editor, but does not support other editors beyond the LSP protocol.

This LSP feature has been requested in https://github.com/microsoft/language-server-protocol/issues/1150
